rust/rope/src/spans.rs

```diff
@@ -209,3 +224,53 @@ impl<'a, T: Clone + Default> Iterator for SpanIter<'a, T> {
         None
     }
 }
+
+impl<'a, T: Clone + Default> Iterator for SpanBackwardsIter<'a, T> {
+    type Item = (Interval, &'a T);
+
+    fn next(&mut self) -> Option<(Interval, &'a T)> {
+        if self.done {
+            return None;
+        }
+        if let Some((leaf, start_pos)) = self.cursor.get_leaf() {
+            if leaf.spans.is_empty() { return None; }
+            let leaf_start = self.cursor.pos() - start_pos;
+            let span = &leaf.spans[leaf.spans.len() - self.ix - 1];
+            self.ix += 1;
+            if self.ix == leaf.spans.len() {
+                if self.cursor.prev_leaf().is_none() {
+                    self.done = true;
+                }
+                self.ix = 0;
+            }
+            return Some((span.iv.translate(leaf_start), &span.data));
+        }
+        None
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use spans::SpansBuilder;
+    use interval::Interval;
+
+    #[test]
+    fn iter_backwards() {
+        let mut sb = SpansBuilder::new(8);
+        sb.add_span(Interval::new_open_open(1, 2), 1);
+        sb.add_span(Interval::new_open_open(4, 4), 3);
+        sb.add_span(Interval::new_open_open(4, 4), 2);
+        sb.add_span(Interval::new_open_open(3, 5), 4);
```

Comment:
This test is dubious because it violates the precondition of `add_span` (sorted start order).